### PR TITLE
Add failure reasons for Moneybookers (Skrill) integration

### PR DESCRIPTION
**Pull request by dhalai**
When a payment fails Skrill will post the failed reason code to the merchant/shopping cart status_url. The mapped failure reason should be displayed in the orders section of the Shopify
backend.

The most recent documentation can be found on https://www.skrill.com/fileadmin/content/pdf/Skrill_Quick_Checkout_Guide.pdf

### CODE REVIEW

File: lib/offsite_payments/integrations/moneybookers.rb [hunk 1]
```diff
@@ -174,23 +214,32 @@ def secret
...
-        # Acknowledge the transaction to MoneyBooker. This method has to be called after a new
-        # apc arrives. It will verify that all the information we received is correct and will return a
-        # ok or a fail. The secret (second credential) has to be provided in the parameter :credential2
-        # when instantiating the Notification object.
+        # Acknowledge the transaction to MoneyBooker.
+        # This method has to be called after a new apc arrives.
+        # It will verify that all the information we received is correct
+        # and will return a ok or a fail.
+        # The secret (second credential) has to be provided in the parameter
+        # :credential2 when instantiating the Notification object.
```
Comment (elfassy) [inline lib/offsite_payments/integrations/moneybookers.rb]: max line width is set at 120 char for this project, no need to modify the comments

File: lib/offsite_payments/integrations/moneybookers.rb [hunk 2]
```diff
@@ -174,23 +214,32 @@ def secret
...
+        #    else
+        #      ... log possible hacking attempt ...
+        #    end
+        #  end
         def acknowledge(authcode = nil)
-          fields = [merchant_id, item_id, Digest::MD5.hexdigest(secret.to_s).upcase, merchant_amount, merchant_currency, status_code].join
+          fields = [
+            merchant_id, item_id, Digest::MD5.hexdigest(secret.to_s).upcase,
+            merchant_amount, merchant_currency, status_code
+          ].join
```
Comment (elfassy) [inline lib/offsite_payments/integrations/moneybookers.rb]: thanks for cleaning up the code. Please keep the esthetic changes in a separate commit

File: test/unit/integrations/moneybookers/moneybookers_notification_test.rb [hunk 3]
```diff
@@ -59,8 +59,15 @@ def test_acknowledge
     assert notification.acknowledge
   end
 
+  def test_failure_reasons
+    Moneybookers::Notification::FAILED_REASON_CODES.each do |code, reason|
+      notification = Moneybookers::Notification.new(http_raw_data.sub(/failed_reason_code=0/, "failed_reason_code=#{code}"), :credential2 => 'secret')
+      assert_equal reason, notification.failure_reason
+    end
+  end
```
Comment (elfassy) [inline test/unit/integrations/moneybookers/moneybookers_notification_test.rb]: please add a test specifically for a code that is not defined in the `FAILED_REASON_CODES`

**Comment by dhalai** [comment]
@elfassy Thanks for reviewing! Done.

**Comment by pi3r** [comment]
Hey @dhalai, sorry for the late feedback here but who is supposed to call `failure_reason`? Is it intended for the shopify integration?

**Comment by dhalai** [comment]
Hey @pi3r, I suppose Shopify should call this method for showing failed reason in the orders section.

**Comment by sergey-alekseev** [comment]
Anyone?

**Comment by pi3r** [comment]
@sergey-alekseev We don't have a logic in Shopify to display an error when an offsite fails. `failure_reason` doesn't fit our api. You can migrate your integration to our Hosted Payment SDK which supports a `message` value. https://docs.google.com/forms/d/e/1FAIpQLSe7aB9HkDQ6z3OgNYanFUUFMofntNlsCmNfwT_pmzfD9Jqi4A/viewform

**Comment by sergey-alekseev** [comment]
@pi3r thanks for your reply.

We'll wait for [Katie MacDonald](https://www.linkedin.com/in/katie-macdonald-7a1366104/) to help with this.

**Comment by dhalai** [comment]
@pi3r So, it's impossible to show the failure reason using offsite payments integration, is it?

**Comment by pi3r** [comment]
@dhalai Offsites work by redirecting buyers to your platform. I think the assumption is you will show the error message on your side.

**Comment by pi3r** [comment]
@dhalai Is this intended to be used on the redirection or when you send us a callback? When we receive a callback, and the status is `failed`, if the notification object responds to the `message` method then it will show this message when you redirect the buyer to the checkout page.

**Comment by dhalai** [comment]
@pi3r Thank you for your answer. We don't want to show this information to the end customers. It should be shown only in the merchant admin area.  

admin_abondoned_checkouts_status

I suppose we can't use the `message` method in this case. What should be used instead?

**Comment by pi3r** [comment]
@dhalai We'll make a change to show the return value of the `#message` method to be used in the admin. Can you make the changes on your side (renaming `failure_reason` to `message`)?

**Comment by dhalai** [comment]
@pi3r thank you, done

**Comment by dhalai** [comment]
Thanks for your help, @pi3r . I just wonder if it's already available on the Shopify side. If not, when are you planning to do it?

**Comment by pi3r** [comment]
@dhalai, it has been merged on our side. Please let me know if it works as expected. Thanks!
